### PR TITLE
L10n tests and better error handling

### PR DESCRIPTION
**Pull request by vwoodzell**
Added some unit tests for l10n. Also changed addL10nSubstitutionInner() to address some problems like the one in bug 6386. Added a test that validates the l10n files, and fixed problems revealed by that test.

### CODE REVIEW

**Comment by Bombe** [comment]
Looks good.

File: src/freenet/l10n/BaseL10n.java [hunk 1]
```diff
@@ -591,14 +592,69 @@ public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, St
...
+     * null if an error is encountered.
+     */
+    private List<HTMLNode> tryGetL10nSubstitution(String key, String value,
+            String[] patterns, HTMLNode[] values) {
+        HTMLNode tempNode = new HTMLNode("#");
+        // catch errors caused by bad translation strings
+        try {
+            addL10nSubstitutionInner(tempNode, value, patterns, values);
+        } catch (L10nParseException e) {
+            Logger.error(this, "Error in l10n value \""+value+"\" for "+key+": "+e.getMessage());
```
Comment (bertm) [inline src/freenet/l10n/BaseL10n.java]: There's `Logger.error(Object o, String s, Throwable e)` for logging exceptions (it also logs the stack trace, which might be useful).

File: src/freenet/l10n/BaseL10n.java [hunk 2]
```diff
@@ -591,14 +592,69 @@ public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, St
...
+        }
+        return newContent;
+    }
+    
+    /**
+     * Performs substitutions on a l10n string to generate a list of {@link HTMLNode}s. Returns
+     * null if an error is encountered.
+     */
+    private List<HTMLNode> tryGetL10nSubstitution(String key, String value,
+            String[] patterns, HTMLNode[] values) {
```
Comment (bertm) [inline src/freenet/l10n/BaseL10n.java]: Looks like this method only performs substitution. There's no need to pass the `key` here.

File: src/freenet/l10n/BaseL10n.java [hunk 3]
```diff
@@ -591,14 +592,69 @@ public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, St
...
+            addL10nSubstitutionInner(tempNode, value, patterns, values);
+        } catch (L10nParseException e) {
+            Logger.error(this, "Error in l10n value \""+value+"\" for "+key+": "+e.getMessage());
+            return null;
+        }
+        return tempNode.getChildren();
+    }
 
 	/**
 	 * @see #addL10nSubstitution(HTMLNode, String, String[], HTMLNode[])
```
Comment (bertm) [inline src/freenet/l10n/BaseL10n.java]: This javadoc does not make sense. `addL10nSubstitution` does key lookup, `addL10nSubstitutionInner` does not.

File: src/freenet/l10n/BaseL10n.java [hunk 4]
```diff
@@ -591,14 +592,69 @@ public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, St
 	 * @param values For each entry in the previous array parameter, this array specifies the {@link HTMLNode} with which it shall be replaced. 
 	 */
 	public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, HTMLNode[] values) {
-		String value = getString(key);
-		addL10nSubstitutionInner(node, key, value, patterns, values);
+        List<HTMLNode> newContent = getL10nSubstitution(key, patterns, values);
+        if (newContent != null) {
```
Comment (bertm) [inline src/freenet/l10n/BaseL10n.java]: `newContent` is never `null` by definition of `getL10nSubstitution`.

**Comment by bertm** [comment]
This is not a blocker, but something I'd like to add anyhow.

I find the naming of the methods confusing. You currently have:
`addL10nSubstitution`, which wraps around
`getL10nSubstitution`, which wraps around
`tryGetL10nSubstitution`, which wraps around
`addL10nSubstitutionInner`

Only the former two are specific to l10n (they pass / lookup keys), the latter two are generally applicable pattern substitution methods. Given the length of this chain, it's a bit odd to call `addL10nSubstitutionInner` the inner method of `addL10nSubstitution`.

As a suggestion, the following might be more clear:

`addL10nSubstitution` (public method, probably shouldn't be renamed),
`getSubstitutionByKey`,
`tryGetSubstitutionByValue`,
`addSubstitutionByValue`

**Comment by bertm** [comment]
Looks pretty sane in general, I only had some minor comments. Special thanks for creating the unit tests!

**Comment by vwoodzell** [comment]
Commits have "changed" due only to merging the first two. Other changes are in new commits.

**Comment by Thynix** [comment]
These tests do not pass on my machine. Have I misconfigured something perhaps?

    [junit] Running freenet.l10n.BaseL10nTest
    [junit] Testsuite: freenet.l10n.BaseL10nTest
    [junit] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.095 sec
    [junit] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.095 sec
    [junit] ------------- Standard Error -----------------
    [junit] Could not get resource : ../test/freenet/l10n/freenet.l10n.en.properties
    [junit] Could not get resource : ../test/freenet/l10n/freenet.l10n.de.properties
    [junit] Could not get resource : ../test/freenet/l10n/freenet.l10n.en.properties
    [junit] The default translation for test.substitution hasn't been found!
    [junit] java.lang.Exception
    [junit] 	at freenet.l10n.BaseL10n.getFallbackString(BaseL10n.java:544)
    [junit] 	at freenet.l10n.BaseL10n.access$000(BaseL10n.java:37)
    [junit] 	at freenet.l10n.BaseL10n$L10nStringIterator.next(BaseL10n.java:165)
    [junit] 	at freenet.l10n.BaseL10n$L10nStringIterator.next(BaseL10n.java:139)
    [junit] 	at freenet.l10n.BaseL10n.getHTMLWithSubstitutions(BaseL10n.java:702)
    [junit] 	at freenet.l10n.BaseL10n.addL10nSubstitution(BaseL10n.java:683)
    [junit] 	at freenet.l10n.BaseL10nTest.testAddL10nSubstitutionMissingFallback(BaseL10nTest.java:156)
    [junit] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [junit] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    [junit] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [junit] 	at java.lang.reflect.Method.invoke(Method.java:606)
    [junit] 	at junit.framework.TestCase.runTest(TestCase.java:176)
    [junit] 	at junit.framework.TestCase.runBare(TestCase.java:141)
    [junit] 	at junit.framework.TestResult$1.protect(TestResult.java:122)
    [junit] 	at junit.framework.TestResult.runProtected(TestResult.java:142)
    [junit] 	at junit.framework.TestResult.run(TestResult.java:125)
    [junit] 	at junit.framework.TestCase.run(TestCase.java:129)
    [junit] 	at junit.framework.TestSuite.runTest(TestSuite.java:255)
    [junit] 	at junit.framework.TestSuite.run(TestSuite.java:250)
    [junit] 	at org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.run(JUnitTestRunner.java:535)
    [junit] 	at org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.launch(JUnitTestRunner.java:1182)
    [junit] 	at org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.main(JUnitTestRunner.java:1033)
    [junit] ------------- ---------------- ---------------
    [junit] 
    [junit] Testcase: testAddL10nSubstitutionMissingBrace took 0.026 sec
    [junit] Testcase: testAddL10nSubstitutionMissingFallback took 0.006 sec
    [junit] 	FAILED
    [junit] expected:<[Text with <b>loud</b> string]> but was:<[test.substitution]>
    [junit] junit.framework.ComparisonFailure: expected:<[Text with <b>loud</b> string]> but was:<[test.substitution]>
    [junit] 	at freenet.l10n.BaseL10nTest.testAddL10nSubstitutionMissingFallback(BaseL10nTest.java:159)
    [junit]

**Comment by vwoodzell** [comment]
Ah, my mistake. I was running without "ant clean", and eclipse had previously copied the test properties files into the build dir. I've added a task to the buildfile to copy them.

**Comment by Thynix** [comment]
Turns out these don't pass in Java 6:

    [junit] Testcase: testGetStringOverridden took 0.005 sec
    [junit] 	FAILED
    [junit] expected:<[O]verridden> but was:<[Not o]verridden>
    [junit] junit.framework.ComparisonFailure: expected:<[O]verridden> but was:<[Not o]verridden>
    [junit] 	at freenet.l10n.BaseL10nTest.testGetStringOverridden(BaseL10nTest.java:171)
    [junit] 

I'll look into it. Any ideas why?

**Comment by Thynix** [comment]
When it breaks under Java 6 it's looking for the override file here: "/home/steve/freenet/fred/run/file:/usr/lib/jvm/java-6-openjdk-common/jre/lib/ext/pulse-java.jar!/../test/freenet/l10n/freenet.l10n.en.override.properties". When it works under Java 7 it's looking for "/home/steve/Documents/Coding/freenet/fred/build/main/../test/freenet/l10n/freenet.l10n.en.override.properties". Looks like the classLoaderDir cannot be assumed to be reasonable. I'll see what the other tests with files do.

**Pull request merged**